core/unix/src/TUnixSystem.cxx

```diff
@@ -4545,64 +4545,84 @@ int TUnixSystem::UnixSend(int sock, const void *buffer, int length, int flag)
 ////////////////////////////////////////////////////////////////////////////////
 /// Get shared library search path. Static utility function.
 
-static const char *DynamicPath(const char *newpath = 0, Bool_t reset = kFALSE)
+static const char *DynamicPath(const char *newpath = nullptr, Bool_t reset = kFALSE)
 {
-   static TString dynpath;
-   static Bool_t initialized = kFALSE;
+   static TString dynpath_full;
+   static std::atomic<bool> initialized(kFALSE);
+   static std::atomic<bool> seenCling(kFALSE);
+
+   // If we have not seen Cling but the result has been initialized and gCling
+   // is still nullptr, the result won't change.
+   if (newpath == nullptr && !reset && (seenCling || (initialized && gCling == nullptr)))
+      return dynpath_full;
+
+   R__LOCKGUARD2(gSystemMutex);
+
+   if (newpath) {
+      dynpath_full = newpath;
+      // Don't erase the user given path at the next call.
+      initialized = kTRUE;
+      seenCling = kTRUE;
```

Comment:
Why is `seenCling` `true` here? Does `seenCling` mean "we have had libCling in the dynamic loader's search path"? (If so, consider putting that as documentation comment for `seenCling`?)

Reply:
We do not (otherwise) record whether the path was set automatically or whether it was set explicitly by the user.  If the user set the path explicitly, we should never automatically over-ride the value;  if `seenCling` stayed false, it would tell this routine that at the next call it should update the value ( to insert the `Cling` provided parts).